pkg/kubectl/bind/authenticator/authenticator.go

```diff
@@ -44,113 +45,102 @@ func init() {
 	utilruntime.Must(kubebindv1alpha1.AddToScheme(kubebindSchema))
 }
 
-type Authenticator interface {
-	Endpoint(context.Context) string
-	Execute(context.Context) error
-}
+type LocalhostCallbackAuthenticator struct {
+	port   int
+	server *http.Server
 
-type defaultAuthenticator struct {
-	server     *echo.Echo
-	port       int
-	timeout    time.Duration
-	actionDone chan struct{}
-	action     func(context.Context, schema.GroupVersionKind, runtime.Object) error
+	mu          sync.Mutex // synchronizes mutations of fields below
+	done        chan struct{}
+	response    runtime.Object
+	responseGvk *schema.GroupVersionKind
 }
 
-func NewDefaultAuthenticator(timeout time.Duration, action func(context.Context, schema.GroupVersionKind, runtime.Object) error) (Authenticator, error) {
-	if timeout == 0 {
-		timeout = 5 * time.Second
+func NewLocalhostCallbackAuthenticator() *LocalhostCallbackAuthenticator {
+	return &LocalhostCallbackAuthenticator{
+		done: make(chan struct{}),
 	}
+}
 
-	defaultAuthenticator := &defaultAuthenticator{
-		timeout:    timeout,
-		action:     action,
-		actionDone: make(chan struct{}),
+func (d *LocalhostCallbackAuthenticator) Start() error {
+	if d.server != nil {
+		return errors.New("already started")
 	}
 
-	server := echo.New()
-	server.HideBanner = true
-	server.HidePort = true
-	server.GET("/callback", defaultAuthenticator.actionWrapper())
-	defaultAuthenticator.server = server
-
 	address, err := net.ResolveTCPAddr("tcp", "localhost:0")
 	if err != nil {
-		return nil, err
+		return err
 	}
 
+	mux := http.NewServeMux()
+	mux.HandleFunc("/callback", d.callback)
+	d.server = &http.Server{Handler: mux}
+
 	listener, err := net.ListenTCP("tcp", address)
 	if err != nil {
-		return nil, err
-	}
-
-	if err = listener.Close(); err != nil {
-		return nil, err
+		return err
 	}
+	d.port = listener.Addr().(*net.TCPAddr).Port
 
-	defaultAuthenticator.port = listener.Addr().(*net.TCPAddr).Port
-
-	return defaultAuthenticator, nil
-}
-
-func (d *defaultAuthenticator) Execute(ctx context.Context) error {
 	go func() {
-		time.Sleep(10 * time.Minute)
-		d.server.Server.Close()
+		d.server.Serve(listener) // nolint: errcheck
 	}()
 
-	go d.server.Start(net.JoinHostPort("localhost", strconv.Itoa(d.port))) // nolint:errcheck
-
-	select {
-	case <-d.actionDone:
-		return nil
-	case <-ctx.Done():
-		return ctx.Err()
-	}
+	return nil
 }
 
-func (d *defaultAuthenticator) Endpoint(context.Context) string {
+// Endpoint returns the URL this server is listening to.
+// Start() must be called prior to this.
+func (d *LocalhostCallbackAuthenticator) Endpoint(context.Context) string {
```

Comment:
why exactly the context? Is that some interface we have to follow?

Reply:
leftover, it was unsued, let me remove